src/mesh/coordinates.cxx

```diff
@@ -119,13 +119,15 @@ Field2D interpolateAndExtrapolate(const Field2D& f, CELL_LOC location,
     }
   }
 
-  // Set corner guard cells
-  for (int i = 0; i < localmesh->xstart; i++) {
-    for (int j = 0; j < localmesh->ystart; j++) {
-      result(i, j) = BoutNaN;
-      result(i, localmesh->LocalNy - 1 - j) = BoutNaN;
-      result(localmesh->LocalNx - 1 - i, j) = BoutNaN;
-      result(localmesh->LocalNx - 1 - i, localmesh->LocalNy - 1 - j) = BoutNaN;
+  if (not localmesh->include_corner_cells) {
```

Comment:
What is the point?
If they are not included, they shouldn't be used.
If you think they are invalid, then you should set them to NaN in all cases.
These seems to me like this is causing unnecessary breakage.
Especially if this is generated from a expression, like `1.0`, there is absolutely nothing wrong with the corner cells, and setting them to NaN has caused issues for me several times in the past.

Reply:
My issue is that in the [coords3d](https://github.com/boutproject/BOUT-dev/tree/coord3d_merged_master) branch I removed the setting to NaN, so that dz is constant if it is generated from a constant expression. Another option would be to extrapolate them if we extrapolate in x and y, however that would be more work ...

Reply:
Could wrap setting them in `#if CHECK > 0` (level to be debated)?

What do you do if dz is not generated from a constant expression? And do you need dz in the corners?

At some we're also going to add guards and boundaries for z, so this may get even more fun.

Reply:
Wrapping in `CHECK` is probably useful anyway, `> 0` is probably fine because it is only used in the init. However, that does not solve this for coords3d.

I check whether `dz` (or `zlength()`) is constant. Currently the check is for `RGN_ALL` as then no further checks are needed, what the actual used domain is, and the first element is used.

In the case of a non-constant `dz`, any of the routines that assume a constant `dz`/`zlength()` cannot be used, so `dz` doesn't need to be constant, and corner cells are no more important then in any other case.

It is probably best to check in all cases what domain is used, and only check constantness for this domain, rather then `RGN_ALL`, but `dz(0,0)` is so much easier then `dz[dz.getRegion(...).begin()[0]]`

To do things right, I think we should:
 * extrapolate into corners, if they are to be used, otherwise leave them. The motivation to not set them to NaN is that they might be generated from a grid or input, which already includes guards and corners. We also do not break boundaries, if they are not extrapolated.
 * use appropriate regions in coord3d

Reply:
> It is probably best to check in all cases what domain is used, and only check constantness for this domain, rather then RGN_ALL, but dz(0,0) is so much easier then dz[dz.getRegion(...).begin()[0]]

Might be wise to wrap this up in something like `constant_dz(const string& region)` that throws if it's not constant over that region.

> extrapolate into corners, if they are to be used, otherwise leave them

Is that not what is done in this change? This flag doesn't say whether or not to use the corners, just whether or not they are communicated.

We could of course mandate that grids for v5 provide values in the boundaries, which removes all this extrapolation stuff. That might just be moving the problem up a level.

Reply:
> > It is probably best to check in all cases what domain is used, and only check constantness for this domain, rather then RGN_ALL, but dz(0,0) is so much easier then dz[dz.getRegion(...).begin()[0]]
> 
> Might be wise to wrap this up in something like `constant_dz(const string& region)` that throws if it's not constant over that region.

:+1: will do

> > extrapolate into corners, if they are to be used, otherwise leave them
> 
> Is that not what is done in this change? This flag doesn't say whether or not to use the corners, just whether or not they are communicated.

I do not see any corner extrapolation. Either they are set to NaN (if `! include_corner_cells`) and otherwise left as they are. With the methods from the non-uniform boundary stuff, it shouldn't be that hard to extend this into 2D/3D to allow extrapolation into corners.

> We could of course mandate that grids for v5 provide values in the boundaries, which removes all this extrapolation stuff. That might just be moving the problem up a level.

Certainly would be easier for the initialization. Also the extrapolation would only need to be done once, at grid creation time. Writing it as script might be easier, because we don't have to think at all possible cases at once.

Also with the extrapolation, I feel uncertain on whether we should include the grid spacing. For dz, dy, dz that is non-trivial, but might be more accurate for the other ones ...

Reply:
Sorry, all the corner-cell handling is rather awful. I probably need more comments/documentation...

If `include_corner_cells == true`, we should be extrapolating into the corners. This PR modifies the boundary regions so that if `include_corner_cells == true` then the y-boundaries include the corner cells. As the y-boundaries are added after the x-boundaries, the `for (auto& bndry : localmesh->getBoundaries()) {` loop above should first extrapolate into the x-boundaries, then into the y-boundaries, including extrapolating into corner cells using values from x-boundary cells.

Reply:
I'd be happy to set the corners to NaN only when we are extrapolating into some boundary cells. Otherwise there is correct data in the corner cells, which there is no reason to overwrite. Keeping it can occasionally be useful: for example when using staggered grids, there is a case when we want to invert a Laplacian at `Ny+1`, and the non-uniform grid correction breaks unless there is valid data in the corner cells.